### PR TITLE
feat: Reputation: CNS-1005 - Reputation pairing score

### PR DESCRIPTION
**Pull request by oren-lava**
# Description

Closes: #XXXX

In this PR I've implemented the reputation pairing score mechanism. After aggregating QoS score over an epoch, every epoch start we update the QoS score for all providers with the epoch QoS score. After that, we group all the providers with the same chain ID and cluster and compare the QoS scores between them. According to the rank, we convert the QoS score to a reputation pairing score which ranges between 0.5-2. This can be used to affect the pairing chance of providers.

Note, the reputation pairing score is fully implemented but disabled. This will be enabled in the future after further analysis (mainly dependent on data collected from raw QoS excellence reports).

---

## Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [ ] read the [contribution guide](https://github.com/lavanet/lava/blob/main/CONTRIBUTING.md)
* [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title, you can find examples of the prefixes below:
    
* [ ] confirmed `!` in the type prefix if API or client breaking change
* [ ] targeted the `main` branch
* [ ] provided a link to the relevant issue or specification
* [ ] reviewed "Files changed" and left comments if necessary
* [ ] included the necessary unit and integration tests
* [ ] updated the relevant documentation or specification, including comments for [documenting Go code](https://blog.golang.org/godoc)
* [ ] confirmed all CI checks have passed

## Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

* [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed all author checklist items have been addressed
* [ ] reviewed state machine logic, API design and naming, documentation is accurate, tests and test coverage

### CODE REVIEW

File: x/pairing/keeper/reputation.go [hunk 1]
```diff
@@ -118,6 +115,241 @@ func (k Keeper) UpdateReputationEpochQosScore(ctx sdk.Context, chainID string, c
...
+			utils.LavaFormatError("setReputationPairingScoreByBenchmark: invalid provider score", fmt.Errorf("score is negative"),
+				utils.LogAttr("chain_id", chainID),
+				utils.LogAttr("cluster", cluster),
+				utils.LogAttr("provider", providerScore.Provider),
+				utils.LogAttr("score", score.String()),
+			)
+		}
+
+		scaledScore := types.MinReputationPairingScore
+		if score.IsZero() || score.LTE(benchmark) {
```
Comment (omerlavanet) [inline x/pairing/keeper/reputation.go:334]: let's explain that since the benchmark is a very high percentile we give perfect score to all providers which keep up with this

File: x/pairing/keeper/reputation.go [hunk 2]
```diff
@@ -118,6 +115,241 @@ func (k Keeper) UpdateReputationEpochQosScore(ctx sdk.Context, chainID string, c
...
+		reputation, err := iter.Value()
+		if err != nil {
+			return nil, utils.LavaFormatError("updateReputationsScores: failed to get reputation from iterator", err,
+				utils.LogAttr("chain_id", chainID),
+				utils.LogAttr("cluster", cluster),
+				utils.LogAttr("provider", provider),
+			)
+		}
+
+		// apply time decay on current score and add the epoch score (which is reset right after)
```
Comment (omerlavanet) [inline x/pairing/keeper/reputation.go:203]: in case no new information was added (epochScore is ZeroQosScore) skip this update

File: x/pairing/keeper/reputation.go [hunk 3]
```diff
@@ -118,6 +115,241 @@ func (k Keeper) UpdateReputationEpochQosScore(ctx sdk.Context, chainID string, c
 	k.SetReputation(ctx, chainID, cluster, provider, r)
```
Comment (omerlavanet) [inline x/pairing/keeper/reputation.go:115]: if we update r.TimeLastUpdated = ctx.BlockTime().UTC().Unix()
this means that every update to the reputation will reset the decay without taking into account
new information in the epochScore should not change the TimeStamp of the existing score in terms of decay
imagine existing score is X for 1 year, now i updated epoch score, timestamp is being set to now, and now i come to the epoch start function, there is 0 decay, we also need to add a test that checks the decay behaviour

**Review by omerlavanet — DISMISSED** [review]
very good code you have come a long way, a few edge cases and decay fixes

File: x/pairing/keeper/reputation.go [hunk 4]
```diff
@@ -118,6 +115,241 @@ func (k Keeper) UpdateReputationEpochQosScore(ctx sdk.Context, chainID string, c
...
+
+// UpdateReputationQosScore updates all the reputations on epoch start with the epoch score aggregated over the epoch
+func (k Keeper) UpdateReputationQosScore(ctx sdk.Context) {
+	// scores is a map of "chainID cluster" -> stakeProviderScores
+	// it will be used to compare providers QoS scores within the same chain ID and cluster and determine
+	// the providers' reputation pairing score.
+	// note, the map is already sorted by QoS score in ascending order.
+	scores, err := k.UpdateReputationsForEpochStart(ctx)
+	if err != nil {
+		panic(utils.LavaFormatError("UpdateReputationQosScore: could not update providers QoS scores", err))
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/pairing/keeper/reputation.go]: done. see https://github.com/lavanet/lava/commit/0b0e7193701728b2ede6ee92b6544be76cfca8fb

File: x/pairing/keeper/reputation.go [hunk 5]
```diff
@@ -118,6 +115,241 @@ func (k Keeper) UpdateReputationEpochQosScore(ctx sdk.Context, chainID string, c
...
+	// iterate over providers QoS scores with the same chain ID and cluster
+	for _, chainCluster := range keys {
+		stakeProvidersScore := scores[chainCluster]
+		split := strings.Split(chainCluster, " ")
+		chainID, cluster := split[0], split[1]
+
+		// get benchmark score value
+		benchmark, err := k.GetBenchmarkReputationScore(stakeProvidersScore)
+		if err != nil {
+			panic(utils.LavaFormatError("UpdateReputationQosScore: could not get benchmark QoS score", err))
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/pairing/keeper/reputation.go]: done. see https://github.com/lavanet/lava/commit/0b0e7193701728b2ede6ee92b6544be76cfca8fb
I can't propagate the error since this function is run from BeginBlock

File: x/pairing/keeper/reputation.go [hunk 6]
```diff
@@ -118,6 +115,241 @@ func (k Keeper) UpdateReputationEpochQosScore(ctx sdk.Context, chainID string, c
...
+		}
+
+		// set reputation pairing score by the benchmark
+		err = k.setReputationPairingScoreByBenchmark(ctx, chainID, cluster, benchmark, stakeProvidersScore.ProviderScores)
+		if err != nil {
+			panic(utils.LavaFormatError("UpdateReputationQosScore: could not set repuatation pairing scores", err))
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/pairing/keeper/reputation.go]: done. see https://github.com/lavanet/lava/commit/0b0e7193701728b2ede6ee92b6544be76cfca8fb

File: x/pairing/keeper/reputation.go [hunk 7]
```diff
@@ -118,6 +115,241 @@ func (k Keeper) UpdateReputationEpochQosScore(ctx sdk.Context, chainID string, c
...
+			providerScores.ProviderScores = []ProviderQosScore{{Provider: provider, Score: reputation.Score, Stake: reputation.Stake}}
+			providerScores.TotalStake = reputation.Stake
+		} else {
+			providerScores.ProviderScores = append(providerScores.ProviderScores, ProviderQosScore{Provider: provider, Score: reputation.Score, Stake: reputation.Stake})
+			providerScores.TotalStake = providerScores.TotalStake.Add(reputation.Stake)
+		}
+		scores[chainID+" "+cluster] = providerScores
+	}
+
+	sortProviderScores(scores)
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/pairing/keeper/reputation.go:248]: done. see https://github.com/lavanet/lava/commit/094ee32996d49feb4cfe2a9f51f23b2c4964b90e

File: x/pairing/keeper/reputation.go [hunk 8]
```diff
@@ -118,6 +115,241 @@ func (k Keeper) UpdateReputationEpochQosScore(ctx sdk.Context, chainID string, c
...
+	}
+	sort.Strings(keys)
+
+	// iterate over providers QoS scores with the same chain ID and cluster
+	for _, chainCluster := range keys {
+		stakeProvidersScore := scores[chainCluster]
+		split := strings.Split(chainCluster, " ")
+		chainID, cluster := split[0], split[1]
+
+		// get benchmark score value
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/pairing/keeper/reputation.go]: done. see https://github.com/lavanet/lava/commit/094ee32996d49feb4cfe2a9f51f23b2c4964b90e

File: x/pairing/keeper/reputation.go [hunk 9]
```diff
@@ -118,6 +115,241 @@ func (k Keeper) UpdateReputationEpochQosScore(ctx sdk.Context, chainID string, c
...
+		if err != nil {
+			return utils.LavaFormatError("setReputationPairingScoreByBenchmark: cannot resolve provider score", err,
+				utils.LogAttr("chain_id", chainID),
+				utils.LogAttr("cluster", cluster),
+				utils.LogAttr("provider", providerScore.Provider),
+			)
+		}
+
+		if score.IsNegative() {
+			utils.LavaFormatError("setReputationPairingScoreByBenchmark: invalid provider score", fmt.Errorf("score is negative"),
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/pairing/keeper/reputation.go]: done. see https://github.com/lavanet/lava/commit/094ee32996d49feb4cfe2a9f51f23b2c4964b90e

File: x/pairing/keeper/reputation.go (hunk 1 above)
Comment (oren-lava) [reply x/pairing/keeper/reputation.go:352]: done. see https://github.com/lavanet/lava/commit/094ee32996d49feb4cfe2a9f51f23b2c4964b90e

File: x/pairing/keeper/reputation.go (hunk 2 above)
Comment (oren-lava) [reply x/pairing/keeper/reputation.go:217]: done. see https://github.com/lavanet/lava/commit/094ee32996d49feb4cfe2a9f51f23b2c4964b90e

File: x/pairing/keeper/reputation.go [hunk 10]
```diff
@@ -118,6 +115,241 @@ func (k Keeper) UpdateReputationEpochQosScore(ctx sdk.Context, chainID string, c
...
+	keys := []string{}
+	for key := range scores {
+		keys = append(keys, key)
+	}
+	sort.Strings(keys)
+
+	// iterate over providers QoS scores with the same chain ID and cluster
+	for _, chainCluster := range keys {
+		stakeProvidersScore := scores[chainCluster]
+		split := strings.Split(chainCluster, " ")
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/pairing/keeper/reputation.go]: done. see https://github.com/lavanet/lava/commit/0989d8df387017b25d6fe2ccf0588a8c0e192ef6

File: x/pairing/keeper/reputation.go (hunk 3 above)
Comment (oren-lava) [reply x/pairing/keeper/reputation.go:113]: done. see https://github.com/lavanet/lava/commit/ee2cf55eebf53a68afaeb921ed4b65905afb08be

File: x/pairing/types/reputation.go [hunk 11]
```diff
@@ -39,7 +48,78 @@ func (r Reputation) ShouldTruncate(stabilizationPeriod int64, currentTime int64)
 	return r.CreationTime+stabilizationPeriod < currentTime
 }
 
+// calcDecayFactor calculates the appropriate decay factor for a reputation from the time it was last updated
+// the decay factor is: exp(-timeSinceLastUpdate/halfLifeFactor)
+func (r Reputation) calcDecayFactor(halfLifeFactor int64, currentTime int64) math.LegacyDec {
+	if halfLifeFactor <= 0 {
```
Comment (Yaroms) [inline x/pairing/types/reputation.go:54]: shouldnt it be uint64 instead?

File: x/pairing/types/reputation.go [hunk 12]
```diff
@@ -39,7 +48,78 @@ func (r Reputation) ShouldTruncate(stabilizationPeriod int64, currentTime int64)
...
+	if timeSinceLastUpdate < 0 {
+		utils.LavaFormatError("calcDecayFactor: calculate reputation decay factor failed, invalid reputation",
+			fmt.Errorf("reputation last update time is larger than current time"),
+			utils.LogAttr("current_time", currentTime),
+			utils.LogAttr("reputation_time_last_updated", r.TimeLastUpdated),
+		)
+		return math.LegacyZeroDec()
+	}
+
+	exponent := float64(timeSinceLastUpdate / halfLifeFactor)
```
Comment (Yaroms) [inline x/pairing/types/reputation.go:72]: we dont use floats on the blockchain

File: x/pairing/keeper/scores/reputation_req.go [hunk 13]
```diff
@@ -0,0 +1,46 @@
+package scores
```
Comment (Yaroms) [inline x/pairing/keeper/scores/reputation_req.go:1]: delete the qos_req file

Comment (oren-lava) [reply x/pairing/keeper/scores/reputation_req.go:1]: done. see https://github.com/lavanet/lava/commit/5073c3860aece3abc12152434b3a1000d8a6a630

File: x/pairing/keeper/reputation.go [hunk 14]
```diff
@@ -118,11 +109,265 @@ func (k Keeper) UpdateReputationEpochQosScore(ctx sdk.Context, chainID string, c
...
+	Stake    sdk.Coin
+}
+
+type StakeProviderScores struct {
+	ProviderScores []ProviderQosScore
+	TotalStake     sdk.Coin
+}
+
+// UpdateReputationQosScore updates all the reputations on epoch start with the epoch score aggregated over the epoch
+func (k Keeper) UpdateReputationQosScore(ctx sdk.Context) {
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/pairing/keeper/reputation.go]: done. see https://github.com/lavanet/lava/commit/5073c3860aece3abc12152434b3a1000d8a6a630

File: x/pairing/types/reputation.go (hunk 11 above)
Comment (oren-lava) [reply x/pairing/types/reputation.go:54]: done. see https://github.com/lavanet/lava/commit/23033489cfdb0424df81bab6e3c42b2ccef5a0ca

File: x/pairing/keeper/keeper.go [hunk 15]
```diff
@@ -130,6 +130,10 @@ func (k Keeper) Logger(ctx sdk.Context) log.Logger {
 
 func (k Keeper) BeginBlock(ctx sdk.Context) {
 	if k.epochStorageKeeper.IsEpochStart(ctx) {
+		// reset pairing query cache every epoch
+		*k.pairingQueryCache = map[string][]epochstoragetypes.StakeEntry{}
```
Comment (Yaroms) [inline x/pairing/keeper/keeper.go:134]: didnt we delete this?

Comment (oren-lava) [reply x/pairing/keeper/keeper.go]: done. see https://github.com/lavanet/lava/commit/87fb3405266d441a56787a656792a486969d2983

File: x/pairing/types/reputation.go (hunk 12 above)
Comment (oren-lava) [reply x/pairing/types/reputation.go]: done. see https://github.com/lavanet/lava/commit/b919b1e234f77dedf72fce32839360619937e4eb

**Review by Yaroms — APPROVED** [review]
